### ISSUE TITLE
training_epoch_end is not computed at the end of epoch

### ISSUE THREAD

**Issue by ysig** [PyTorchLightning/pytorch-lightning#8204]
## 🐛 Bug

Hi, I see a very simple bug in the implementation of `training_epoch_end`.
Precisely it is computed close to the end of the epoch, but not in the end, as in the following:
![image](https://user-images.githubusercontent.com/28439529/123856441-c52ed780-d929-11eb-81a0-24952934f6e3.png)
(Compute prints the inputs that have been seen until then which is batch-size\*3241 (where batch-size 2) and not batch-size\*3729.)

My `training_epoch_end` looks like the following:
```
    def training_epoch_end(self, outputs):
        # We can directly implement the method ".compute()" from the accuracy function
        cer, wer, ser = self.metrics.compute()

        # Save the metric
        self.log('CER', cer, prog_bar=True)
        self.log('WER', wer, prog_bar=True)
        self.log('SER', ser, prog_bar=True)
```

and my  trainer is like:

` trainer = Trainer(gpus=1, callbacks=[checkpoint_callback], auto_select_gpus=True, max_epochs=100, logger=logger)
`

### Expected behavior
`training_epoch_end` should run at the **end** of an epoch.

### Version
'1.3.7post0'

**Comment by carmocca** [comment]
> Precisely it is computed close to the end of the epoch, but not in the end

How are you observing this exactly?

is the `print(COMPUTE!, ...)` statement in your `training_epoch_end` function and you are making this assumption because the progress bar is not full?

Keep in mind that the progress bar includes both the training and validation progress.

> cer, wer, ser = self.metrics.compute()

Unrelated, but you might be interested in upstreaming you Metric to resolve https://github.com/PyTorchLightning/metrics/issues/52 👀

**Comment by ysig** [comment]
Hi,

thank you for your answer.
I don't think it is a problem with my metric.
`print(COMPUTE!, ...)` is called in `self.metrics.compute()` which itself is called on `training_epoch_end`.

> Keep in mind that the progress bar includes both the training and validation progress.
Ah ok, then I was probably confused because of that.
But this is misleading:
It would be better if you updated the tqdm description of the Epoch progress bar from `train` to `val` instead of just adding a nested progress bar on validation. What is the intuition behind such a design choice?

Thanks (you can close the comment)

**Comment by carmocca** [comment]
> It would be better if you updated the tqdm description of the Epoch progress bar from train to val instead of just adding a nested progress bar on validation. What is the intuition behind such a design choice?

That this design fits better when validation runs at any point of the training epoch (and/or multiple times). It allows you to see the full progress of the fit epoch (train + validation)

If you would like to, you can open a PR changing it so the progress bar description is updated when the stage changes
